src/thor/route_action.cc

```diff
@@ -127,6 +128,83 @@ void remove_edges(const GraphId& edge_id, valhalla::Location& loc, GraphReader&
 namespace valhalla {
 namespace thor {
 
+std::string thor_worker_t::expansion(Api& request) {
+  // default the expansion geojson so its easy to add to as we go
+  rapidjson::Document dom;
+  dom.SetObject();
+  rapidjson::Pointer("/type").Set(dom, "FeatureCollection");
+  rapidjson::Pointer("/properties/algorithm").Set(dom, "none");
+  rapidjson::Pointer("/features/0/type").Set(dom, "Feature");
+  rapidjson::Pointer("/features/0/geometry/type").Set(dom, "MultiLineString");
+  rapidjson::Pointer("/features/0/geometry/coordinates").Create(dom).SetArray();
+  rapidjson::Pointer("/features/0/properties/edge_ids").Create(dom).SetArray();
+  rapidjson::Pointer("/features/0/properties/statuses").Create(dom).SetArray();
+
+  // a lambda that the path algorithm can call to add stuff to the dom
+  auto track_expansion = [&dom](baldr::GraphReader& reader, const char* algorithm,
+                                baldr::GraphId edgeid, const char* status, bool full_shape = false) {
+    // full shape might be overkill but meh, its trace
+    const auto* tile = reader.GetGraphTile(edgeid);
+    const auto* edge = tile->directededge(edgeid);
+    auto shape = tile->edgeinfo(edge->edgeinfo_offset()).shape();
+    if (!edge->forward())
+      std::reverse(shape.begin(), shape.end());
+    if (!full_shape && shape.size() > 2)
+      shape.erase(shape.begin() + 1, shape.end() - 1);
+
+    // make the geom
+    auto& a = dom.GetAllocator();
+    auto* coords = rapidjson::Pointer("/features/0/geometry/coordinates").Get(dom);
+    coords->GetArray().PushBack(rapidjson::Value(rapidjson::kArrayType), a);
+    auto& linestring = (*coords)[coords->Size() - 1];
+    for (const auto& p : shape) {
+      linestring.GetArray().PushBack(rapidjson::Value(rapidjson::kArrayType), a);
+      auto point = linestring[linestring.Size() - 1].GetArray();
+      point.PushBack(p.first, a);
+      point.PushBack(p.second, a);
+    }
+
+    // make the properties
+    rapidjson::Pointer("/properties/algorithm").Set(dom, algorithm);
+    rapidjson::Pointer("/features/0/properties/edge_ids")
+        .Get(dom)
+        ->GetArray()
+        .PushBack(static_cast<uint64_t>(edgeid), a);
+    rapidjson::Pointer("/features/0/properties/statuses")
+        .Get(dom)
+        ->GetArray()
+        .PushBack(rapidjson::Value{}.SetString(status, a), a);
+  };
```

Comment:
we inject this lambda into the path algorithm and it calls us back with it. here we are building geojson but we could do something else with this information if we wanted.